### PR TITLE
Add phone auth

### PR DESCRIPTION
**Pull request by kroikie**
Update firebase_auth plugin to support phone number authentication.

This PR is a combination of the ideas expressed by @alibitek in [PR](https://github.com/flutter/plugins/pull/329) and @pranjalsatija in [PR](https://github.com/flutter/plugins/pull/484)

Fixes: https://github.com/flutter/flutter/issues/10404

### CODE REVIEW

**Comment by pranjalsatija** [comment]
@kroikie looks good. Sorry about the formatting issue. I'm having some serious compatibility issues with my computer after upgrading to Mojave.

**Comment by pranjalsatija** [comment]
@kroikie Should we also add a corresponding `phoneNumber` field to `FirebaseUser`?

**Comment by kroikie** [comment]
@pranjalsatija good call, added.

**Comment by mmcc007** [comment]
I tested it on an armv7 (32-bit) iPhone and an android simulator and the example provided worked fine. Nice work.

checked out the PR using 
```
git fetch origin pull/606/head:mybranch
git checkout mybranch
```
Note have to add phone number `+1 408-555-6969` and verifier `888888` to whitelist in Google Console if testing provided example on your own Google project (in addition to the usual handling of `google-services.json` and `GoogleService-Info.plist`).

To add it to your own flutter project and see how far you can get, add the following to your `pubspec.yaml`:
```
  firebase_auth:
    git:
      url: git@github.com:flutter/plugins.git
      path: packages/firebase_auth
      ref: 075897e
```
Happy Coding!!

**Comment by hemesh2394** [comment]
Is firebase auth v 0.5.12 is released?
how i can implement to my project?

**Comment by mmcc007** [comment]
@hemesh2394 As of this posting, `firebase auth v 0.5.12` has not yet been released. This Pull Request tracks the process to release. `firebase auth v 0.5.12` is currently considered implemented and must go thru a review process before being released (additional development and review may occur before being released). In the meantime, if you wish to experiment with the new phone feature prior to release, you can use the current (unreleased as of this posting) version by following my suggestion in prior post.

**Comment by saivamsid** [comment]
@mmcc007 how much time will it take for the release of firebase auth v 0.5.12. Thanks

**Comment by gildaswise** [comment]
@mmcc007 How did you manage to get the iOS working? Here it just stopped sending the phone verification today on a real device while the simulator still works.

**Comment by Hixie** [comment]
cc @bparrishMines for review

**Comment by mmcc007** [comment]
@gildaswise I tested it again on an armv7 (32-bit) iPhone and it still works (deleted the user in Google Console -> Authentication). It looks like it will not work without using a normalized phone number (probably conforming to format described in [libphonenumber](https://github.com/googlei18n/libphonenumber)). For example, in my case (in USA) it did not send a verification code if I used the format `14085556969` but did when I used the format `+14085556969`. 

BTW: would be nice if there was a dart implementation of  [libphonenumber](https://github.com/googlei18n/libphonenumber) to catch badly formed phone numbers on input (hint... hint...  ;) )

**Comment by mmcc007** [comment]
@gildaswise I assume you are using your `real iPhone 7+`'s phone number on your `real iPhone 7+`. Try texting (or have someone from another country text) your `real iPhone 7+` using an inbound international format to confirm you are using the correct format for your `real iPhone 7+`

Per [Telephone_numbers_in_Brazil](https://en.wikipedia.org/wiki/Telephone_numbers_in_Brazil#Inbound) I'm guessing the format should look something like `011 55 aa nnnn nnnn`.  My guess is to try it without the spaces, so for example for Rio de Janeiro (area code 21) it would be `011552155555555`. Worth a try.

**Comment by gildaswise** [comment]
@mmcc007 That didn't work, sadly. As I said, it's only failing on the iPhone 7+ while it works normally with this format: `+55 (21) 95555-5555` parsed like this: `+5521955555555` on the Android side or the iOS simulator (SE or 8+)

**Comment by mmcc007** [comment]
@gildaswise Working on a simulator confirms that the code is most likely completely correct. 

If you cannot receive a text on your real device from whichever data center Google is texting your confirmation code (datacenter could be in Brazil, or some other country), then, I'm guessing, either you cannot receive international texts on your real device, or the format of the number is incorrect, or possibly something else. 

Confirm u can receive international texts on your real device and try different phone number formats, I'm guessing it should eventually work.

**Comment by gildaswise** [comment]
@mmcc007 I tried sending the code to the iPhone 7+'s number from my Google Pixel and it worked. I really don't know what is the problem as all printed logs from inputted numbers are matching.

**Comment by mmcc007** [comment]
My guess is that your real device has to have a plan from your telephone company that supports international texting.

**Comment by gildaswise** [comment]
I said that it's receiving the code if I do the process from my Android device or the simultator, but it isn't "sending from itself".

**Comment by mmcc007** [comment]
@gildaswise I assume from your posts that the phone auth of your flutter app works on your `real Android device` but not on your `real iPhone device` (in addition to your `iPhone simulator`). I can confirm I made no special changes for my `real iPhone device`. It worked first time after it worked on my `android simulator`. 

If both your `real iPhone and Android devices` have a plan that supports international texting, then I don't know off the top of my head, why you can receive confirmation codes on your `real Android device` but not on your `real iPhone device`.

My guess is you're pretty close to solving the problem. I don't know if there is much further I can add.

File: packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java [hunk 1]
```diff
@@ -110,12 +117,113 @@ public void onMethodCall(MethodCall call, Result result) {
...
+        break;
       default:
         result.notImplemented();
         break;
     }
   }
 
+  private void handleSignInWithPhoneNumber(MethodCall call, Result result) {
+    Map<String, String> arguments = (Map<String, String>) call.arguments;
+    String verificationId = arguments.get("verificationId");
```
Comment (bparrishMines) [inline packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java]: If `call.arguments` is a map, I believe you can use `call.arguments("verificationId")` instead. Then you could avoid the casting above.

File: packages/firebase_auth/example/lib/main.dart [hunk 2]
```diff
@@ -37,6 +37,10 @@ class MyHomePage extends StatefulWidget {
 
 class _MyHomePageState extends State<MyHomePage> {
   Future<String> _message = new Future<String>.value('');
+  TextEditingController smsCodeController = new TextEditingController();
```
Comment (bparrishMines) [inline packages/firebase_auth/example/lib/main.dart]: nit: add underscore to variables like `_message`

File: packages/firebase_auth/example/lib/main.dart [hunk 3]
```diff
@@ -105,6 +155,31 @@ class _MyHomePageState extends State<MyHomePage> {
...
                 });
               }),
+          new MaterialButton(
+              child: const Text('Test verifyPhoneNumber'),
+              onPressed: () {
+                _testVerifyPhoneNumber();
+              }),
+          new Container(
+            margin: const EdgeInsets.only(
+                top: 8.0, bottom: 8.0, left: 16.0, right: 16.0),
```
Comment (bparrishMines) [inline packages/firebase_auth/example/lib/main.dart]: nit: comma after `right` parameter

**Comment by mmcc007** [comment]
@saltpetreca I don't know of any single place that walks thru the entire set-up for phone auth on flutter.  You can sort of piece it together by looking at phone [auth](https://firebase.google.com/docs/auth/) for ios/android. 

The provided example app does not work out-of-the-box for phone auth (AFAIK by design). 

In my case, I pointed the provided example app to my existing firebase project (with usual handling of `google-services.json` and `GoogleService-Info.plist`)

I then enabled the phone provider (in google console) with the white listed phone number `+1 408-555-6969` (used in provided example app) and validation code `888888`. 

As I recall, in the provided example app, `+1 408-555-6969` is hard coded. After starting up the provided example app, a submit will send the phone number (this would normally result in a validation code via SMS text to your phone, but since white listed, no SMS text is sent). Then enter the white listed validation code `888888`. Submitting again will send the validation code, and u should then see a response security string on screen indicating user is authorized for firebase.

**Comment by saltpetreca** [comment]
Thanks I did the same thing of pointing to my project and updating the plist. Been at it since last time. But it keeps crashing.

Wanted to scrap the example and rewrite it myself but hard to find documentation. 

I am guessing calls to FB are the same through this plugin.

Thanks

On Jul 4, 2018, at 12:00 PM, Maurice McCabe <notifications@github.com> wrote:

@saltpetreca I don't know of any single place that walks thru the entire set-up for phone auth on flutter. You can sort of piece it together by looking at phone auth for ios/android.

The provided example app does not work out-of-the-box for phone auth (AFAIK by design).

In my case, I pointed the provided example app to my existing firebase project (with usual handling of google-services.json and GoogleService-Info.plist)

I then enabled the phone provider (in google console) with the white listed phone number +1 408-555-6969 (used in provided example app) and validation code 888888.

As I recall, in the provided example app, +1 408-555-6969 is hard coded. After starting up the provided example app, a submit will send the phone number (this would normally result in a validation code via SMS text to your phone, but since white listed, no SMS text is sent). Then enter the white listed validation code 888888. Submitting again will send the validation code, and u should then see a response security string on screen indicating user is authorized for firebase.

**Comment by mmcc007** [comment]
@saltpetreca I don't recall having to change any code to get example app to work, just the configuration. However, had to be a particularly careful about what button to press, when to enter validation code, and in what sequence (buttonx->validation code->buttonx). 

As I recall, email auth worked for me via google email provider with this example app.

**Comment by saltpetreca** [comment]
Thanks will give that a try. Does the example app crash on you when for both ios and android. For ios only the anon function works  if I click on anything else it crashes.

On Jul 4, 2018, at 8:25 PM, jasonge511 <notifications@github.com> wrote:

@saltpetreca Replace the signInWithGoogle method in the example app with the signInWithPhoneNumber method exposed in this plugin. You then have to piece together the available properties from the JS implementation of Phone Auth provided in the firebase docs.

Still very early days for this feature, any ETA/priority update? I can help with dartlang package documentation for this feature.

**Comment by AlexandreRoba** [comment]
Hi all,

I'm trying to setup the phone  auth. Unfortunattely when I call the method verifyPhoneNumber The apps crash directly on simulator and on my Phone (Both IOS) with the following error:
```bash
Lost connection to device.
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001053e41e6 __exceptionPreprocess + 294
	1   libobjc.A.dylib                     0x0000000104a79031 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000105459975 +[NSException raise:format:] + 197
	3   Runner                              0x00000001024ab7eb -[FIRPhoneAuthProvider verifyPhoneNumber:UIDelegate:completion:] + 187
	4   Runner                              0x0000000102523761 -[FLTFirebaseAuthPlugin handleMethodCall:result:] + 12849
	5   Flutter                             0x0000000102a499ea __45-[FlutterMethodChannel setMethodCallHandler:]_block_invoke + 118
	6   Flutter                             0x0000000102a61a1a _ZNK5shell21PlatformMessageRouter21HandlePlatfo<…>
```

Any idea what I could Investigate?

**Comment by saltpetreca** [comment]
Try adding below and go a get package see if that helps in your case. 
Have the exact same issue as you still troubleshooting.

firebase_core: "^0.2.4"


On Jul 5, 2018, at 10:55 AM, Alexandre Roba <notifications@github.com> wrote:

Hi all,

I'm trying to setup the phone auth. Unfortunattely when I call the method verifyPhoneNumber The apps crash directly on simulator and on my Phone (Both IOS) with the following error:

Lost connection to device.
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001053e41e6 __exceptionPreprocess + 294
	1   libobjc.A.dylib                     0x0000000104a79031 objc_exception_throw + 48
	2   CoreFoundation                      0x0000000105459975 +[NSException raise:format:] + 197
	3   Runner                              0x00000001024ab7eb -[FIRPhoneAuthProvider verifyPhoneNumber:UIDelegate:completion:] + 187
	4   Runner                              0x0000000102523761 -[FLTFirebaseAuthPlugin handleMethodCall:result:] + 12849
	5   Flutter                             0x0000000102a499ea __45-[FlutterMethodChannel setMethodCallHandler:]_block_invoke + 118
	6   Flutter                             0x0000000102a61a1a _ZNK5shell21PlatformMessageRouter21HandlePlatfo<…>
Any idea what I could Investigate?

**Comment by AlexandreRoba** [comment]
Hi @saltpetreca, thanks for getting back to me. I have those two dart package on my pubspec :
```yaml
firebase_auth:
    git:
      url: git@github.com:flutter/plugins.git
      path: packages/firebase_auth
      ref: 075897e
firebase_core: ^0.2.4
```

For info I'm also using the anonymous authentication and this works fine. It is only crashing for the phone authentication. So I guess hte pods should be ok but not sure. I went through the Firebase Silent notification setup but was not sure this was needed...

**Comment by AlexandreRoba** [comment]
Hi @saltpetreca, I found out why. The silent notification is also not working on my phone... Meaning it is trying to start the reCaptcha verification. I did not setup that part properly. Once I did (Adding the Url type then It started to work properly...) I need now to find out why is the silent notification not working...

**Comment by saltpetreca** [comment]
Good job man! can you post you changes. I have been at it for the last 2 days couldn't fig it out. Does it work for you both on ios and android? 

On Jul 5, 2018, at 12:45 PM, Alexandre Roba <notifications@github.com> wrote:

Hi @saltpetreca, I found out why. The silent notification is also not working on my phone... Meaning it is trying to start the reCaptcha verification. I did not setup that part properly. Once I did (Adding the Url type then It started to work properly...) I need now to find out why is the silent notification not working...

**Comment by AlexandreRoba** [comment]
Hi @saltpetreca,

Enable the recaptcha as explained here on this [link](https://firebase.google.com/docs/auth/ios/phone-auth#set-up-recaptcha-verification)

And in order to enable the silent push notification to avoid the captch I add to enable the Background refresh following this [link](https://www.raywenderlich.com/143128/background-modes-tutorial-getting-started)

I currently focused only the authentication on IOs. I will start working on the android probably tomorrow. I will let you know if I manage to have it working.

Cheers.

**Comment by saltpetreca** [comment]
Thanks giving it a try. Let me know how the android part paned out. Crashes on me on startup.

On Jul 5, 2018, at 2:14 PM, Alexandre Roba <notifications@github.com> wrote:

Hi @saltpetreca,

Enable the recaptcha as explained here on this link

And in order to enable the silent push notification to avoid the captch I add to enable the Background refresh following this link

I currently focused only the authentication on IOs. I will start working on the android probably tomorrow. I will let you know if I manage to have it working.

Cheers.

**Comment by gildaswise** [comment]
Hey! Passing by to say that what I was missing on my setup was APNs configuration (as explained [here](https://firebase.google.com/docs/auth/ios/phone-auth#start-receiving-silent-notifications)) and that I just published a Medium tutorial for this: https://medium.com/@gildaswise/flutter-adding-sign-in-with-google-and-phone-authentication-to-your-app-69f681518f9b

**Comment by mmcc007** [comment]
From a clean checkout of this PR I did the following to get the example app working for phone auth. I confirmed phone auth worked on both the ios and android emulators and also an iPhone device.

(some... or all... of these changes are required for google sign-in too... if that service provider is enabled in console. So was able to test google sign-in works too)

checked out the PR using
```
git fetch origin pull/606/head:mybranch
git checkout mybranch
```
The following files were changed manually in `mybranch`  (other changes to these files occur during build):
```
 M android/app/build.gradle <-- see diffs
 M android/app/google-services.json <-- overwrite with your file
 M ios/Flutter/AppFrameworkInfo.plist <-- see diffs
 M ios/Flutter/Debug.xcconfig <-- see diffs... to fix a build problem on ios emulator
 M ios/Runner.xcodeproj/project.pbxproj <-- see diffs
 M ios/Runner/GoogleService-Info.plist <-- overwrite with your file
 M ios/Runner/Info.plist <-- see diffs... reCAPTCHA verification
?? android/key.properties <-- link to a private key ... see below
```
Instructions for generating the private key required for android auth are [here](https://developers.google.com/android/guides/client-auth).
Then create the following file:
`android/key.properties`
```
storePassword=yourstorepassword
keyPassword=yourkeypassword
keyAlias=key
storeFile=/pathtoprivatekey/key.jks
```
On the server side, enable the phone auth provider in the console. To test example app you will need to add the white-listed phone number `+1 408-555-6969` and verification code `888888`.

When the app is running:
1.  click on `Test verifyPhoneNumber`
2. enter SMS code `888888`
3. click on `Test signinWithPhoneNumber`

You should then see the following on screen:
```
signInWithPhoneNumber succeeded: FirebaseUser({providerId: firebase, uid: xxxxxxxxxxxxxxxxxxxxxxxxxxxx, phoneNumber: +14085556969, isAnonymous: false, isEmailVerified: false, providerData: [{providerId: firebase, uid: xxxxxxxxxxxxxxxxxxxxxxxxxxxx, phoneNumber: +14085556969}]})
```

To see the manual changes to existing files in `mybranch` paste the following diffs to https://diffy.org (I took out the diffs I figured were generated during build... includes handling of reCAPTCHA verification for iOS mentioned by @AlexandreRoba and described [here](https://firebase.google.com/docs/auth/ios/phone-auth#set-up-recaptcha-verification)):
```
diff --git a/packages/firebase_auth/example/android/app/build.gradle b/packages/firebase_auth/example/android/app/build.gradle
index 696849c..43179c5 100755
--- a/packages/firebase_auth/example/android/app/build.gradle
+++ b/packages/firebase_auth/example/android/app/build.gradle
@@ -23,7 +23,7 @@ android {
 
     defaultConfig {
         minSdkVersion 16
-        applicationId "io.flutter.plugins.firebaseauthexample"
+        applicationId "com.yourcompany.yourapp"
         testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
     }
 
diff --git a/packages/firebase_auth/example/ios/Flutter/Debug.xcconfig b/packages/firebase_auth/example/ios/Flutter/Debug.xcconfig
index 9803018..626b4ec 100755
--- a/packages/firebase_auth/example/ios/Flutter/Debug.xcconfig
+++ b/packages/firebase_auth/example/ios/Flutter/Debug.xcconfig
@@ -1,2 +1,3 @@
 #include "Generated.xcconfig"
 #include "Pods/Target Support Files/Pods-Runner/Pods-Runner.debug.xcconfig"
+FLUTTER_BUILD_MODE=debug
diff --git a/packages/firebase_auth/example/ios/Runner.xcodeproj/project.pbxproj b/packages/firebase_auth/example/ios/Runner.xcodeproj/project.pbxproj
old mode 100755
new mode 100644
index eebc126..f31c4ff
--- a/packages/firebase_auth/example/ios/Runner.xcodeproj/project.pbxproj
+++ b/packages/firebase_auth/example/ios/Runner.xcodeproj/project.pbxproj
@@ -453,7 +454,7 @@
 					"$(inherited)",
 					"$(PROJECT_DIR)/Flutter",
 				);
-				PRODUCT_BUNDLE_IDENTIFIER = io.flutter.plugins.firebaseAuthExample;
+				PRODUCT_BUNDLE_IDENTIFIER = com.yourcompany.yourapp;
 				PRODUCT_NAME = "$(TARGET_NAME)";
 			};
 			name = Debug;
@@ -476,7 +477,7 @@
 					"$(inherited)",
 					"$(PROJECT_DIR)/Flutter",
 				);
-				PRODUCT_BUNDLE_IDENTIFIER = io.flutter.plugins.firebaseAuthExample;
+				PRODUCT_BUNDLE_IDENTIFIER = com.yourcompany.yourapp;
 				PRODUCT_NAME = "$(TARGET_NAME)";
 			};
 			name = Release;
diff --git a/packages/firebase_auth/example/ios/Runner/Info.plist b/packages/firebase_auth/example/ios/Runner/Info.plist
index 873ea0e..46e2c2b 100755
--- a/packages/firebase_auth/example/ios/Runner/Info.plist
+++ b/packages/firebase_auth/example/ios/Runner/Info.plist
@@ -18,6 +18,17 @@
 	<string>1.0</string>
 	<key>CFBundleSignature</key>
 	<string>????</string>
+	<key>CFBundleURLTypes</key>
+	<array>
+		<dict>
+			<key>CFBundleTypeRole</key>
+			<string>Editor</string>
+			<key>CFBundleURLSchemes</key>
+			<array>
+				<string>com.googleusercontent.apps.your project number-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx</string>
+			</array>
+		</dict>
+	</array>
 	<key>CFBundleVersion</key>
 	<string>1</string>
 	<key>LSRequiresIPhoneOS</key>
@@ -45,19 +56,5 @@
 	</array>
 	<key>UIViewControllerBasedStatusBarAppearance</key>
 	<false/>
-	<!-- Google Sign-in Section -->
-	<key>CFBundleURLTypes</key>
-	<array>
-		<dict>
-			<key>CFBundleTypeRole</key>
-			<string>Editor</string>
-			<key>CFBundleURLSchemes</key>
-			<array>
-				<!-- Copied from GoogleServices-Info.plist key REVERSE_CLIENT_ID -->
-				<string>com.googleusercontent.apps.466218294969-5mdmh6opkqd87a5m81nm83u4eqd3bvgk</string>
-			</array>
-		</dict>
-	</array>
-	<!-- End of the Google Sign-in Section -->
 </dict>
 </plist>

```

Hope this helps.

**Comment by saltpetreca** [comment]
@gildaswise & @mmcc007 THANK YOU I got it to work on IOS but android still crashes on start up emulator and real device. Trying to figure it out. Any suggestion will be appreciated. 
@gildaswise Read your article. Stellar job in writing it up in such detail. Really appreciate it.
One question. when phone auth becomes part of the firebase_auth package. what will happen if we use this PR. Do we have to redo it per the firebase_auth package or this PR will get merged in a way that we can still keep using it. 

Thanks

**Comment by pranjalsatija** [comment]
Once phone auth gets merged into the official repo (hopefully soon, this PR has been open for a month now), you’ll be able to include the `firebase_auth` package in your `pubspec.yaml` as usual, without any need for referencing a specific commit. AFAIK, you *should* still be able to reference the PR like the article mentions, but there won’t be a need to.

**Comment by pranjalsatija** [comment]
Once phone auth gets merged into the official repo (hopefully soon, this PR has been open for a month now), you’ll be able to include the `firebase_auth` package in your `pubspec.yaml` as usual, without any need for referencing a specific commit. AFAIK, you *should* still be able to reference the PR like the article mentions, but there won’t be a need to.

**Comment by gildaswise** [comment]
@saltpetreca On Android it'll only crash if you haven't included your `google-services.json` file in `android/app/`! Also, try running `flutter packages upgrade` too. Thanks for the compliments!

**Comment by PrimeNexes** [comment]
If anyone has successfully implemented a basic phone auth please do share a rough document to implement it .

**Comment by saltpetreca** [comment]
@gildaswise. Double check the .json file its there. redownloaded and added again still the same issue. Getting below warining but it builds and as soon as it starts on the emulator it crashes.

Launching lib/main.dart on Android SDK built for x86 in debug mode...
registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)
Built build/app/outputs/apk/debug/app-debug.apk.

**Comment by saltpetreca** [comment]
@PrimeNexes take a look at an article written by @gildaswise  https://medium.com/@gildaswise/flutter-adding-sign-in-with-google-and-phone-authentication-to-your-app-69f681518f9b

And above comments on this page by @mmcc007 

They will give you a solid starting point.

**Comment by gildaswise** [comment]
@saltpetreca Sadly these warnings have nothing to do with your problem. Can you test on a real device? Anything else appears on Android Studio's logcat?

**Comment by mmcc007** [comment]
@saltpetreca did u set-up the keystore containing the SHA-1 of your signing certificate for android and link your app to it? (I updated instructions above to correct use of terminology). Auth will not work for android without this.

**Comment by saltpetreca** [comment]
Finally got it to work on Android. I made a typo on the package name. it was off by a letter. But don't know why it didn't give any errors or warning in the debug log.  
Also @mmcc007/@gildaswise does this plugin have all the option as compared to using firebase_auth for native android or IOS. For instances implementing persistence login. In the examples as soon as you close the app or it goes to background. the login expire and the phone auth kicks in again.

**Comment by pranjalsatija** [comment]
@saltpetreca Persisting the verification ID in the middle of the authentication flow on iOS and Android is the developer's responsibility. See [this link](https://firebase.google.com/docs/auth/ios/phone-auth#send-a-verification-code-to-the-users-phone) for more (look at step 3). If you want to do this in Flutter, you can use [`shared_preferences`](https://pub.dartlang.org/packages/shared_preferences).

**Comment by saltpetreca** [comment]
@pranjalsatija Thanks

**Comment by mmcc007** [comment]
@saltpetreca I have not seen that behaviour on my app. After the login is complete the logged state appears to be persisted on the device unless explicitly changed using [signOut](https://github.com/flutter/plugins/blob/master/packages/firebase_auth/lib/firebase_auth.dart#L270).

**Comment by CasperPas** [comment]
I tried on my real phone and almost everything works. `verificationCompleted` callback always returns `null` user. Input code manually still works, but auto auth is not. My phone runs Android 8.1, and I granted `SMS_READ` permission. Any ideas?

**Comment by mmcc007** [comment]
@CasperPas You do not need to grant `SMS_READ` permission on android (or iOS).  If not using a white-listed number (and on a real device) the server will send an SMS text using existing SMS messaging already available on your phone.

I completed the steps described in [my comment](https://github.com/flutter/plugins/pull/606#issuecomment-402878854) above to get phone auth to work on iOS and android (using the example and the white-listed number). Once phone auth is working in example app, you should have enough to make it work on your app on real devices receiving real SMS text messages.

**Comment by CasperPas** [comment]
@mmcc007 yes, SMS works fine. I mean, auto retrieval/instant verification does not work. Is it because of Firebase itself?
My code looks like this

```dart
Future<void> _verifyPhoneNumber() async {
  await _auth.verifyPhoneNumber(
    phoneNumber: phoneNumber,
    timeout: _timeOut,
    codeSent: codeSent,
    codeAutoRetrievalTimeout: codeAutoRetrievalTimeout,
    verificationCompleted: verificationCompleted,
    verificationFailed: verificationFailed);
}

verificationCompleted(FirebaseUser user) async {
  print("onVerificationCompleted, user: $user");
}
```
Then the log shows
> I/flutter ( 2496): onVerificationCompleted, user: null

**Comment by mmcc007** [comment]
@CasperPas the setup for auth (both phone and email) requires many configuration steps... Getting the example app to work requires no change to code, only configuration changes. 

If not already done, I recommend getting auth to work with example app first... to confirm you have made all the configuration changes correctly... before moving-on to getting auth to work with your app... avoids fiddling with code if the problem is with the configuration.

**Comment by mmcc007** [comment]
@degroote22 I provided steps for checking out this PR in [this comment](https://github.com/flutter/plugins/pull/606#issuecomment-402878854) ...as well as configuration, building and testing of example. Let me know if they don't work. 😊

**Comment by AlexandreRoba** [comment]
Hi all, I'm having trouble on my build server building the application since I have specifed this PR in the pubspec. Anyone can guide me in the rigt direction?

pubspec contains in the dependencies:
```yaml
dependencies:
  flutter:
    sdk: flutter
  flutter_localizations:
    sdk: flutter

  # The following adds the Cupertino Icons font to your application.
  # Use with the CupertinoIcons class for iOS style icons.
  cupertino_icons: ^0.1.2
  fluro: ^1.3.1
  flutter_map: ^0.0.6
  latlong: ^0.4.5
  rxdart: ^0.18.1
  date_format: ^1.0.4
  firebase_core: ^0.2.4
    #firebase_auth: ^0.5.12
  firebase_auth:
    git:
      url: git@github.com:flutter/plugins.git
      path: packages/firebase_auth
      ref: 075897e
  location: ^1.3.3
```

I can flutter packages get on my laptop but on the build server I'm getting the following error:
```bash
Running "flutter packages get" in mobilet_app...
Could not find a file named "packages/firebase_auth/pubspec.yaml" in git@github.com:flutter/plugins.git 075897eaec21dffd238ddfaa544ae7b773fc3b29.
```

I have the same version of flutter installed on both:

this is what I have on the server:
```bash
Config:
- Version: 0.5.6-dev
- WorkingDir: /Users/vagrant/git/sweetnest_app
- Commands: [build ios -t lib/main_impl.dart]
ANDROID_HOME environment variable not defined, skipping Android SDK setup.
Extracting Flutter SDK to /Users/vagrant/Library/flutter
Executing Flutter command: build ios -t lib/main_impl.dart
  ╔════════════════════════════════════════════════════════════════════════════╗
  ║                 Welcome to Flutter! - https://flutter.io                   ║
  ║                                                                            ║
  ║ The Flutter tool anonymously reports feature usage statistics and crash    ║
  ║ reports to Google in order to help Google contribute improvements to       ║
  ║ Flutter over time.                                                         ║
  ║                                                                            ║
  ║ Read about data we send with crash reports:                                ║
  ║ https://github.com/flutter/flutter/wiki/Flutter-CLI-crash-reporting        ║
  ║                                                                            ║
  ║ See Google's privacy policy:                                               ║
  ║ https://www.google.com/intl/en/policies/privacy/                           ║
  ║                                                                            ║
  ║ Use "flutter config --no-analytics" to disable analytics and crash         ║
  ║ reporting.                                                                 ║
  ╚════════════════════════════════════════════════════════════════════════════╝
  
Running "flutter packages get" in sweetnest_app...
Could not find a file named "packages/firebase_auth/pubspec.yaml" in git@github.com:flutter/plugins.git 075897eaec21dffd238ddfaa544ae7b773fc3b29.
---- Log transcript ----
FINE: Pub 2.0.0-dev.63.0.flutter-4c9689c1d2
MSG : Resolving dependencies...
SLVR: fact: sweetnest_app is 0.0.0
SLVR: derived: sweetnest_app
SLVR: fact: sweetnest_app depends on date_format ^1.0.4
SLVR: fact: sweetnest_app depends on fluro ^1.3.1
SLVR: fact: sweetnest_app depends on location ^1.3.3
SLVR: fact: sweetnest_app depends on flutter_localizations any from sdk
SLVR: fact: sweetnest_app depends on flutter any from sdk
SLVR: fact: sweetnest_app depends on firebase_core ^0.2.4
SLVR: fact: sweetnest_app depends on firebase_auth from git
SLVR: fact: sweetnest_app depends on cupertino_icons ^0.1.2
SLVR: fact: sweetnest_app depends on rxdart ^0.18.1
SLVR: fact: sweetnest_app depends on latlong ^0.4.5
SLVR: fact: sweetnest_app depends on flutter_map ^0.0.6
SLVR: fact: sweetnest_app depends on flutter_test any from sdk
SLVR: fact: sweetnest_app depends on flutter_launcher_icons ^0.3.2
```

And on my laptop:
```bash
➜ sweetnest/sweetnest_app ➤ ba5edf5|develop ✓
± flutter doctor                                                                                                                                           [8h]
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, v0.5.7-pre.122, on Mac OS X 10.13.5 17F77, locale en-BE)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
[✓] Android Studio (version 3.1)
[✓] IntelliJ IDEA Ultimate Edition (version 2018.1.6)
[✓] VS Code (version 1.25.0)
[✓] Connected devices (1 available)

• No issues found!
➜ sweetnest/sweetnest_app ➤ ba5edf5|develop ✓
± flutter packages get                                                                                                                                     [8h]
Running "flutter packages get" in sweetnest_app...           0.9s
```

Any help appreciated :)

**Comment by kueckermann** [comment]
@AlexandreRoba Have you tried using https instead of ssh? It fixed some fetch issues for me.
`firebase_auth:
    git:
      url: https://github.com/flutter/plugins.git
      path: packages/firebase_auth
      ref: 000cd4c7f8b9eba0feef714a08ff8a13f0f81990`

**Comment by AlexandreRoba** [comment]
Hello @kueckermann thanks for the tip. It solved the issue. Now why it works on my mac and not in a docker container is mistery :). Cheers, Alex.

File: packages/firebase_auth/example/lib/main.dart (hunk 2 above)
Comment (kroikie) [reply packages/firebase_auth/example/lib/main.dart]: Done

File: packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java (hunk 1 above)
Comment (kroikie) [reply packages/firebase_auth/android/src/main/java/io/flutter/plugins/firebaseauth/FirebaseAuthPlugin.java]: The cast to `Map<String, String>` would still be necessary so I'll leave it as is.

File: packages/firebase_auth/example/lib/main.dart (hunk 3 above)
Comment (kroikie) [reply packages/firebase_auth/example/lib/main.dart]: Done

**Comment by kroikie** [comment]
@xzminx Thanks for reporting this, I was able to reproduce and I will submit a fix in another PR.

**Comment by kroikie** [comment]
@xzminx fix for this issue is available in [PR](https://github.com/flutter/plugins/pull/670)

**Comment by probuse** [comment]
@kueckermann I am getting 
`fatal: bad object 000cd4c7f8b9eba0feef714a08ff8a13f0f81990`
What should i use for the ref part?
When i use ssh i get:
```
Permission denied (publickey).
fatal: Could not read from remote repository.
```
How can i add the dependency for the PR in my flutter project, 
Thank you

**Comment by kroikie** [comment]
@probuse Phone Auth has been published in the [firebase_auth plugin](https://pub.dartlang.org/packages/firebase_auth).

**Comment by saltpetreca** [comment]
@gildaswise If you have a chance can you please update your medium article with 5.0.15. Thanks

**Comment by saltpetreca** [comment]
@gildaswise Thanks Appreciate

**Comment by padmanabanSampath** [comment]
@gildaswise & @kroikie 
How to persist the user data for certain period in phone auth, even if the user closes the app?

**Comment by CasperPas** [comment]
@padmanabanSampath you can use *shared_preferences* plugin

**Comment by padmanabanSampath** [comment]
@CasperPas thanks for your recommendation, any example code snippet will be useful, regarding phone auth token or uid persistence and  checking whether the user has signed in before or not, and adding expiry to the shared preference firebase user storage

**Comment by shriharip** [comment]
I am sorry if this is not the right place to raise this issue again. Just opened the issue on flutter (https://github.com/flutter/flutter/issues/22903 ) but was looking for any help!?.

Is a google or any other auth necessary for phone auth or can it be done stand alone. If it can be done stand alone what is the above error message.

˜Thanks

**Pull request merged**